src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs

```diff
@@ -97,97 +113,123 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
         public async Task SetSession(CollaborationSession session)
         {
             _session = session;
-            var roots = await session.ListRootsAsync(CancellationToken.None).ConfigureAwait(false);
-            if (roots.Length > 0)
-            {
-                _remoteRootPath = session.ConvertSharedUriToLocalPath(roots[0]);
-                _remoteRootPath = _remoteRootPath.Substring(0, _remoteRootPath.Length - 1);
-                var lastSlash = _remoteRootPath.LastIndexOf('\\');
-                _externalPath = _remoteRootPath.Substring(0, lastSlash + 1);
-                _externalPath += "~external";
-            }
             IsRemoteSession = true;
+
+            // Files could have been opened before the collaboration session was fully initialized.
+            // So get the roots and update remote paths for all open files.
+            await UpdatePathsToRemoteFilesAsync().ConfigureAwait(false);
+
             session.RemoteServicesChanged += (object sender, RemoteServicesChangedEventArgs e) =>
             {
                 _remoteDiagnosticListTable.UpdateWorkspaceDiagnosticsPresent(_session.RemoteServiceNames.Contains("workspaceDiagnostics"));
             };
         }
 
-        public void EndSession()
+        public bool IsExternalLocalUri(string localPath)
+            => _registeredExternalPaths.Any(externalPath => localPath.StartsWith(externalPath) && localPath.Length > externalPath.Length + 1);
```

Comment:
They in fact do not have trailing.  I'm not sure why as yes it would consider the not really path part of it.  I'll make the change to just keep the trailing as I don't know why we got rid of them.